### ISSUE TITLE
Gunicorn gthread worker not releasing memory

### ISSUE THREAD

**Issue by pratapkishorevarma** [benoitc/gunicorn#2065]
When using the gthread worker for a simple flask app like below, the memory of the worker process gradually increases and even after all the processing is done the memory used by the worker never comes down.

When other worker types like gevent are used, it seems not to consume that much memory

In the below app I am loading the csv file of 10 mb size. After some time the memory usage crosses more than one GB when gthread is used.

Environment:
```
Linux(Centos 7.5, RHEL 8)
python 3.6
gunicorn config
workers = 1
worker_class = "gthread"
threads = 30
timeout = 1800 
keepalive = 2
```

```python
from flask import Flask
app = Flask(__name__)

import pandas as pd
from concurrent.futures import ThreadPoolExecutor

pool = ThreadPoolExecutor(max_workers=10)

@app.route('/test')
def test():
    pool.submit(read_data)
    return 'Read data'

def read_data():
    df = pd.read_csv("data.csv")
    df = None
    print("read data")
    return
```

**Comment by pratapkishorevarma** [comment]
I see that in the latest gthread.py code there is a function get_thread_pool which can be used to overwrite the pool used. I don't see this is included in the 19.9.0 version. When can we expect the next version of gunicorn?

**Comment by tilgovi** [comment]
> When can we expect the next version of gunicorn?

We do not have the resources right now to commit to any timeline.

CPython itself does not release memory immediately. It can keep memory reserved to improve the speed of future allocations. Only when there is memory pressure from the OS will Python free these objects, I think.

Are you finding that this causes swapping or other problems? What is the total memory on the machine?

If you can reproduce a memory leak in the threaded worker with a simple example, that would constitute a bug that should be fixed.

**Comment by benoitc** [comment]
> I see that in the latest gthread.py code there is a function get_thread_pool which can be used to overwrite the pool used. I don't see this is included in the 19.9.0 version. 

Can you describe your usage for?

**Comment by MeteHanC** [comment]
Hey, so I am making a load test to an app which is served with gunicorn. About the load test, I am using 50 threads in parallel and making around 15K requests per minute. 

> 
> Are you finding that this causes swapping or other problems? What is the total memory on the machine?

   Yes, I got 64 GB of memory, with flask(and it's default router), memory consumption is around 3-4 GB all the time (I run the test for around an hour). But with using Starlette and serving it with Gunicorn memory consumption increases continuously and eventually it causes swapping. And even If I kill the gunicorn process, sub processes do not get killed and memory deallocation does not happen. So it forces me to reboot the server. @benoitc  @tilgovi

**Comment by tilgovi** [comment]
@MeteHanC your issue may be different. The original report is about the `gthread` worker, but if you're using starlette I would expect you are using the `uvicorn` worker?

**Comment by tilgovi** [comment]
@pratapkishorevarma can you confirm that this is a leak in `gunicorn` by removing the CSV load?

If it's not reproducible without loading the CSV, then there is probably no bug in the worker, since your example does not do anything with the dataframe other than instantiate it and then release it.

You may also be seeing normal behavior of Python: http://effbot.org/pyfaq/why-doesnt-python-release-the-memory-when-i-delete-a-large-object.htm

**Comment by tilgovi** [comment]
> And even If I kill the gunicorn process, sub processes do not get killed and memory deallocation does not happen. So it forces me to reboot the server.

What subprocesses? Gunicorn workers or additional subprocess that you spawn? How do you spawn them? If you kill the master Gunicorn process, the workers should exit unless something you are initializing in your application is overriding Gunicorn signal handling.

**Comment by MeteHanC** [comment]
> @MeteHanC  I would expect you are using the `uvicorn` worker?

Yes, I am using uvicorn worker. 





> > And even If I kill the gunicorn process, sub processes do not get killed and memory deallocation does not happen. So it forces me to reboot the server.
> 
> What subprocesses? Gunicorn workers or additional subprocess that you spawn? How do you spawn them? If you kill the master Gunicorn process, the workers should exit unless something you are initializing in your application is overriding Gunicorn signal handling.

Gunicorn workers, I am not spawning any additional subprocesses. I am running gunicorn with 48 workers and 2 threads. Under the load test, it keeps spawning new processes/tasks and if I don't stop the load test it runs out of memory, and If I stop the load test and kill the main gunicorn process, it does not release the memory and does not kill the sub gunicorn workers as well. @tilgovi

**Comment by benoitc** [comment]
@MeteHanC I do,n't understand that part:

>  Under the load test, it keeps spawning new processes/tasks

Can ou elaborate what you mean there? ADo you have any logs or a simple way to reproduce your issue?

**Comment by sergicollado** [comment]
> > @MeteHanC  I would expect you are using the `uvicorn` worker?
> 
> Yes, I am using uvicorn worker.
> 
> > > And even If I kill the gunicorn process, sub processes do not get killed and memory deallocation does not happen. So it forces me to reboot the server.
> > 
> > 
> > What subprocesses? Gunicorn workers or additional subprocess that you spawn? How do you spawn them? If you kill the master Gunicorn process, the workers should exit unless something you are initializing in your application is overriding Gunicorn signal handling.
> 
> Gunicorn workers, I am not spawning any additional subprocesses. I am running gunicorn with 48 workers and 2 threads. Under the load test, it keeps spawning new processes/tasks and if I don't stop the load test it runs out of memory, and If I stop the load test and kill the main gunicorn process, it does not release the memory and does not kill the sub gunicorn workers as well. @tilgovi

I have same problem in production with 75k conections ... my kubernetes PODs are getting restarted without memory with 100GB Disks.

**Comment by MeteHanC** [comment]
> I have same problem in production with 75k conections ... my kubernetes PODs are getting restarted without memory with 100GB Disks.

Strange thing is I have tried the same load test with using Flask and it's default HTTP router. Still the same behaviour, it does not release the memory and does not kill the tasks as well. I think this could be about Python's own threads and task management.

**Comment by ivanlmj** [comment]
Yup. I observed something similar as well for a Django app that I'm running. Seems like workers are not releasing memory:

`top - 21:43:13 up 62 days, 19:41,  1 user,  load average: 0.00, 0.00, 0.00
Tasks: 120 total,   1 running,  77 sleeping,   0 stopped,   0 zombie
%Cpu0  :  0.0 us,  0.0 sy,  0.0 ni,100.0 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
%Cpu1  :  0.0 us,  0.0 sy,  0.0 ni,100.0 id,  0.0 wa,  0.0 hi,  0.0 si,  0.0 st
KiB Mem :  4037948 total,   716056 free,  3060072 used,   261820 buff/cache
KiB Swap:  2097148 total,  1484448 free,   612700 used.   749924 avail Mem 

  PID USER      PR  NI    VIRT    RES    SHR S  %CPU %MEM     TIME+ COMMAND                                                                                                                                                                                                  
13230 gunicorn  20   0 3262572 1.780g   8148 S   0.0 46.2   4:36.99 /opt/app/venv_app/bin/python3 /opt/app/venv_app/bin/gunicorn --timeout=60 --workers=5 -b unix:/tmp/app_wsgi.sock app.wsgi:application                                
13236 gunicorn  20   0 1257428 375516   4708 S   0.0  9.3   6:11.97 /opt/app/venv_app/bin/python3 /opt/app/venv_app/bin/gunicorn --timeout=60 --workers=5 -b unix:/tmp/app_wsgi.sock app.wsgi:application                                
 2843 gunicorn  20   0 1144420 254960  20512 S   0.0  6.3   1:26.21 /opt/app/venv_app/bin/python3 /opt/app/venv_app/bin/gunicorn --timeout=60 --workers=5 -b unix:/tmp/app_wsgi.sock app.wsgi:application                                
 3480 gunicorn  20   0  850480 202616   8072 S   0.0  5.0   2:06.90 /opt/app/venv_app/bin/python3 /opt/app/venv_app/bin/gunicorn --timeout=60 --workers=5 -b unix:/tmp/app_wsgi.sock app.wsgi:application                                
 3561 gunicorn  20   0  599184 160444   4932 S   0.0  4.0   0:45.77 /opt/app/venv_app/bin/python3 /opt/app/venv_app/bin/gunicorn --timeout=60 --workers=5 -b unix:/tmp/app_wsgi.sock app.wsgi:application                   `

**Comment by MeteHanC** [comment]
> Yup. I observed something similar as well for a Django app that I'm running. Seems like workers are not releasing memory. I using the default worker (sync):

Are you using multiprocessing libray ? Using multiprocessing.Pool was the issue for me, I made a workaround, now it is deallocating memory as it is supposed to. However, this bug still  remains

**Comment by pratapkishorevarma** [comment]
@tilgovi Sorry for the very late reply. I somehow missed the messages.
I agree with you that this is might be a python problem and not gunicorn specific issue.
Looks like the python thread pool doesn't release the memory as the threads in the pool are reused.

**Comment by tilgovi** [comment]
I don't believe this ticket is actionable without steps to reproduce it. If someone can reproduce an issue and share a repository that would be most helpful.

**Comment by benoitc** [comment]
old issue without any activity. closing it. feel free to open *a new ticket*  if needed.
